src/librustc_trans/back/link.rs

```diff
@@ -665,6 +662,26 @@ fn link_natively(sess: &Session,
         let mut out = output.stderr.clone();
         out.extend(&output.stdout);
         let out = String::from_utf8_lossy(&out);
+
+        // Check to see if the link failed with "unrecognized command line option:
+        // '-no-pie'". If so, reperform the link step without the -no-pie option. This
+        // is safe because if the linker doesn't support -no-pie then it should not
+        // default to linking executables as pie. Different versions of gcc seem to
+        // use different quotes in the error message so don't check for them.
+        if out.contains("unrecognized command line option") && out.contains("-no-pie") {
```

Comment:
To help prevent an accidental infinite loop, could this also have a clause  that the `cmd` has an argument of `-no-pie` in it?